code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -212,31 +279,100 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+    peersResources <-
+        atomically $ HMap.elems <$> readTVar (ztPeers ^. pivConnecting)
+    forM_ peersResources $ \(_,sock,adapter) ->
+        Z.close sock >> adapterRelease adapter
+
+-- Connects/disconnects peers at request.
+changePeers :: MonadIO m => ZTGlobalEnv -> ZTNetCliEnv -> ZTUpdatePeersReq -> m ()
+changePeers ZTGlobalEnv{..} ZTNetCliEnv{..} req = liftIO $ do
+    curTime <- getCurrentTimeMs
+
+    -- We pre-create sockets to use them in the STM. The number may be
+    -- slightly more than the real number used (due to the possibly
+    -- incorrect request), but we'll cleanup right after.
+    sockets <- replicateM (length $ req ^. uprAdd) (Z.socket ztContext Z.Dealer)
+    adapters <- mapM newSocketAdapter sockets
 
-changePeers :: MonadIO m => ZTNetCliEnv -> ZTUpdatePeersReq -> m ()
-changePeers ZTNetCliEnv{..} req = liftIO $ do
-    curTime <- getCurrentTimeMS
     (toConnect,toDisconnect) <- atomically $ do
-        peers <- readTVar ztPeers
-        let (toAdd,toDel) = applyUpdatePeers peers req
-        let peers' = (peers `Set.difference` toDel) `Set.union` toAdd
-        writeTVar ztPeers $! peers'
-        -- 2 seconds heuristical delay before heartbeating worker
-        -- starts noticing these nodes.
-        let initHbs = HBState hbIntervalMin hbLivenessMax (curTime + 2000) False
+        -- Process request
+        (toAdd,toDel) <- applyUpdatePeers ztPeers req
+
+        -- Modify actual variables content.
+        modifyTVar (ztPeers ^. pivConnected) $! \p ->
+            p `HMap.difference` HMap.fromList (map (,()) toDel)
+        let addInfo = zipWith (\s a -> (curTime,s,a)) sockets adapters
+        modifyTVar (ztPeers ^. pivConnecting) $! \p ->
+            p `HMap.union` HMap.fromList (toAdd `zip` addInfo)
+
+        -- Remove them from the heartbeating worker.
         modifyTVar ztHeartbeatInfo $
-            foldr (.) id (map (\nId -> at nId .~ Nothing) (Set.toList toDel)) .
-            foldr (.) id (map (\nId -> at nId .~ Just initHbs) (Set.toList toAdd))
+            foldr (.) id (map (\nId -> at nId .~ Nothing) toDel)
+
+        -- Initialise related heartbeating context
         pure (toAdd,toDel)
+
     forM_ toDisconnect $ \z -> do
-        ztLog ztCliLogging Info $ "changePeers: disconnecting " <> show z
         Z.disconnect ztCliBack $ ztNodeIdRouter z
         Z.disconnect ztCliSub $ ztNodeIdPub z
-    forM_ toConnect $ \z -> do
-        ztLog ztCliLogging Info $ "changePeers: connecting " <> show z
-        Z.connect ztCliBack $ ztNodeIdRouter z
-        Z.connect ztCliSub $ ztNodeIdPub z
-
+    unless (null toDisconnect) $
+        ztLog ztCliLogging Info $ "changePeers: disconnected " <> show toDisconnect
+    unless (null toConnect) $ do
+        -- We connect all dealer sockets and send a connection request
+        -- message.
+        forM_ (toConnect `zip` sockets) $ \(z,d) -> do
+            Z.connect d $ ztNodeIdRouter z
+            Z.sendMulti d $ NE.fromList ["","getid"]
```

Comment:
How can I make it as a pattern and a function alias at the same time? e.g. I now have `pattern GetId` which I can match on, but I can't use it in this particular `zendMulti d ["", GetId]`. If there's a solution I'll refactor this.

Reply:
Update: I have done it straightforwardly without pattern synonyms.

Reply:
Pattern synonyms can be symmetric, thus I thought you could use them as value alias at the same time. Something like
```
pattern ConnectionReq cId t = [cId,"",t]
pattern Msg cId t msgT msg = cId:"":t:msgT:msg
```

Reply:
@volhovm I'd like you to respond on this (ack or nack), then I approve.

> Update: I have done it straightforwardly without pattern synonyms.
I'm actually more concerned about overall message format being duplicated, not only tags.

Reply:
I think that somehow replacing this will only obscure reader's understanding of how ZMQ sockets work. Yes it looks kind of unusual and strange, but if you are acquainted with ZMQ itself it becomes obvious and even (imho) more straightforward rather than using any kind of wrappers. So I'd like to leave it as it is. Though your suggestion about internal messages was good :)

Reply:
Okay, thank you :+1: